File: problems/010-all-or-nothing-output-collection/solutions/python/main.py
import sys

t = sys.stdin.buffer.read().split()
it = iter(t)
N, Q, U = map(int, (next(it), next(it), next(it)))
a = sorted((next(it), int(next(it)), int(next(it))) for _ in range(N))
pre = [0]
mismatch = N
for i, (_, m, x) in enumerate(a):
    pre.append(pre[-1] + m)
    if mismatch == N and m != x:
        mismatch = i
k = 0
out = []
for _ in range(Q):
    b = int(next(it))
    if b < U:
        out.append("ERR QUOTA -")
        continue
    cap = b - U
    while k < N and pre[k + 1] <= cap:
        k += 1
    if k < mismatch:
        out.append("ERR QUOTA " + a[k][0].decode())
    elif mismatch < N:
        out.append("ERR MISMATCH " + a[mismatch][0].decode())
    elif k < N:
        out.append("ERR QUOTA " + a[k][0].decode())
    else:
        out.append(f"OK {N} {U+pre[N]}")
print("\n".join(out))

File: problems/010-all-or-nothing-output-collection/solutions/rust/main.rs
use std::io::{self, Read};
struct File {
    p: String,
    m: u64,
    a: u64,
}
fn main() {
    let mut s = String::new();
    io::stdin().read_to_string(&mut s).unwrap();
    let mut t = s.split_whitespace();
    let n: usize = t.next().unwrap().parse().unwrap();
    let q: usize = t.next().unwrap().parse().unwrap();
    let u: u64 = t.next().unwrap().parse().unwrap();
    let mut f = Vec::with_capacity(n);
    for _ in 0..n {
        f.push(File {
            p: t.next().unwrap().to_string(),
            m: t.next().unwrap().parse().unwrap(),
            a: t.next().unwrap().parse().unwrap(),
        });
    }
    f.sort_by(|a, b| a.p.cmp(&b.p));
    let mut pre = vec![0u64; n + 1];
    let mut mismatch = n;
    for i in 0..n {
        pre[i + 1] = pre[i] + f[i].m;
        if mismatch == n && f[i].m != f[i].a {
            mismatch = i;
        }
    }
    let mut k = 0usize;
    let mut out = String::new();
    for _ in 0..q {
        let b: u64 = t.next().unwrap().parse().unwrap();
        if b < u {
            out += "ERR QUOTA -\n";
            continue;
        }
        let cap = b - u;
        while k < n && pre[k + 1] <= cap {
            k += 1
        }
        if k < mismatch {
            out += &format!("ERR QUOTA {}\n", f[k].p)
        } else if mismatch < n {
            out += &format!("ERR MISMATCH {}\n", f[mismatch].p)
        } else if k < n {
            out += &format!("ERR QUOTA {}\n", f[k].p)
        } else {
            out += &format!("OK {} {}\n", n, u + pre[n])
        }
    }
    print!("{}", out);
}
